Jetpack::dns_prefetch converted to using wp_resource_hints()

willstocks opened the pull request:
See #8090 and #10330

NOTE: This PR started out as #10349, looking solely at Comment Likes, then looking at all `Jetpack::dns_prefetch` functions to potentially convert to `wp_resource_hints()` and therefore deprecate the Jetpack specific function, per #8090 (comment). Now, however, it makes more sense to keep the `Jetpack::dns_prefetch` function in place, but have that pass the values over to `wp_resource_hints()` as opposed to manually managing the `<link>`'s via a custom function.

### Changes proposed in this Pull Request:
The Comment Likes feature current registers dns-prefetch resources via Jetpack's own function. Since the creation of this feature/function, wp_resource_hints() has been created, that allows a user/plugin/theme to register resource hints for use in browsers.
The `Jetpack::dns_prefetch` function is in use in approximately 6 places throughout Jetpack:

- \modules\comments.php
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/comments.php#L33
- \modules\comment-likes.php
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/comment-likes.php#L14
- \modules\likes.php
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/likes.php#L14
- \modules\masterbar\masterbar.php
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/masterbar/masterbar.php#L34
- \modules\photon-cdn.php
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/photon-cdn.php#L17
- \modules\photon.php
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/photon.php#L16
- \modules\shortcodes\videopress.php
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/shortcodes/videopress.php#L11

###Testing instructions:

1. Enable each feature individually
2. Browse to page with that feature on
3. Check page source code with your browser's dev console. You should see <link rel="dns-prefetch" href="//url-relevant-to-feature"> is now in your head section

###Proposed changelog entry for your changes:
Utilize wp_resource_hints() function to register dns-prefetch URLs, rather than a custom function.

willstocks committed:
Jetpack::dns_prefetch converted to using wp_resource_hints()

kraftbj commented on class.jetpack.php:
```diff
@@ -6933,21 +6933,24 @@ public static function build_raw_urls( $url ) {
-			echo "\r\n";
-			foreach ( $prefetch_urls as $this_prefetch_url ) {
-				printf( "<link rel='dns-prefetch' href='%s'/>\r\n", esc_attr( $this_prefetch_url ) );
+			if( 'dns-prefetch' == $relation_type ) {
```
Where does `$relation_type` come from?

willstocks-tech replied:
I've pushed another commit up, with this value statically declared now. However I don't feel like it's the best approach - it's potentially a value that you could pass back as another parameter (alongside the URLs) for differing types of `<link>`?
Implementing this is a _little_ outisde of my current comfort zone, though I'll keep looking/trying!

willstocks-tech replied:
So, I have now statically declared that in the latest commit, however I don't feel like that's the best approach as it's inflexible. Say in the future you want to pass `preconnect` or another `<link>` type, we could extend the `Jetpack::dns_prefetch` to pass that variable back?
Bit outside of my current comfort zone, but I'll give it a whirl :)

kraftbj commented on class.jetpack.php:
```diff
@@ -6933,21 +6933,24 @@ public static function build_raw_urls( $url ) {
+					$urls[] = strtolower( untrailingslashit( preg_replace( '#^https?://#i', '//', $this_new_url ) ) );
+				}
+				$urls = array_unique( $urls );
+			return $urls;
```
I don't believe this `return` is needed. The function has a dual-purpose—if a new URL is passed to it, it will add it to a static variable to keep all of them stored. If no new URL is passed, it outputs them. In this case, a new URL is being passed, so the function, as originally written, shouldn't output anything.

willstocks-tech replied:
Removed in new commit :)

kraftbj commented on class.jetpack.php:
```diff
@@ -6933,21 +6933,24 @@ public static function build_raw_urls( $url ) {
 		}
 	}
-
+	add_filter( 'wp_resource_hints', 'dns_prefetch', 10, 2 );
```
This can't be placed here (within a class but not within a function). You'll see that other filters are added in the constructor: https://github.com/Automattic/jetpack/blob/master/class.jetpack.php#L508

Also, the filter needs to use the right callable. `dns_prefetch` would just look for a regular function by that name, not a method function within a class. https://github.com/Automattic/jetpack/blob/master/class.jetpack.php#L541 is an example of how to ask the hook to use a class method.

willstocks-tech replied:
Removed in new commit and I will push into `class.jetpack.php` similarly to the line you've mentioned

willstocks-tech replied:
OK, line now moved :)

kraftbj reviewed (changes requested):
Thanks for your work on this! There are a few issues that need to be addressed before the PR is testable. Overall, I like the approach of keeping the Jetpack function as a wrapper to ease other developers on using it, keeping the code DRY, and to ensure the URLs are unique.

willstocks committed:
Update class.jetpack.php

Removal of unnecessary `return $urls;`
Removal of incorrectly placed `add_filter( 'wp_resource_hints', 'dns_prefetch', 10, 2 );`
Addition of  `static $relation_type = 'dns-prefetch';`

willstocks committed:
Update class.jetpack.php

Added `add_filter( 'wp_resource_hints', 'dns_prefetch', 10, 2 );` back to file, in correct place

willstocks-tech commented:
A few changes made as requested @kraftbj

jeherve commented on class.jetpack.php:
```diff
@@ -6932,22 +6933,23 @@ public static function build_raw_urls( $url ) {
 	 */
 	public static function dns_prefetch( $new_urls = null ) {
```
Since the `wp_resource_hints` hook accepts 2 parameters, you would have to specify `$relation_type` here, as well as set a default value for it.

jeherve commented on class.jetpack.php:
```diff
@@ -6932,22 +6933,23 @@ public static function build_raw_urls( $url ) {
 	 */
 	public static function dns_prefetch( $new_urls = null ) {
 		static $prefetch_urls = array();
+		static $relation_type = 'dns-prefetch'; /* Should this be a passed param, to allow different link types*/
```
Yes, it would be best.

willstocks-tech replied:
I've just realised, whilst it makes sense to pass this as a parameter for now, if you were to look at using other rel types in the future, you'd probably want to rename the function. 
In theory as well, in it's current state the function could probably be cleaned up to wrap the entire entire thing in the `if ($relation_type == 'dns-prefetch)` rather than having two nested. That way, you only have to maintain the single `if` and adding others would be _fairly_ easy from then on - again with a single `if`.
Hopefully I'm making sense there!

jeherve reviewed (changes requested):
I would recommend making some more changes I highlighted, and then testing your changes on a test site to see how it all works.

willstocks committed:
Update class.jetpack.php

Remove statically declared `$relation_type = 'dns-prefetch'` and pass as variable back via `Jetpack::dns_prefetch` function
Add `$relation_type` as a function parameter
Added missing { } from nested `if` statement within the `elseif`

jeherve commented on class.jetpack.php:
```diff
@@ -6929,25 +6930,27 @@ public static function build_raw_urls( $url ) {
+	 * @param string | $relation_type | resource hint type
 	 */
-	public static function dns_prefetch( $new_urls = null ) {
+	public static function dns_prefetch( $new_urls = null, $relation_type ) {
```
Could you add a default value here for `$relation_type`, to avoid issues when the function is called without a second parameter?

willstocks-tech replied:
Yep, just caught that - commited up with this change, setting a default value of `$relation_type = 'dns-prefetch'`

jeherve replied:
The default value should be all you need; since you have a default `ab98c7244215e0306bee2acdfe57c9cf48f3b8ff` is not necessarily needed.

willstocks-tech replied:
Very true! Can get rid of that commit for now, as noted in another comment, it probably wouldn't make much sense to call `Jetpack::dns_prefetch` if you were wanting to `<link rel="preconnect">` as the function name would throw you off? Overtime you guys might want to look at changing from `Jetpack::dns_prefetch` to something like `Jetpack::relation_types` (probably could come up with a better name!!)

willstocks-tech replied:
@jeherve - I believe I have now reverted that commit: ab98c7244215e0306bee2acdfe57c9cf48f3b8ff

jeherve reviewed (changes requested):
Another small change you'll need to make.

willstocks committed:
Add `, dns-prefetch` to all `Jetpack::dns_prefetch` functions, to pass back as parameter for use with `wp_resource_hints()`

willstocks committed:
Default value for `$relation_type` parameter, to prevent issues if second param is not passed

willstocks committed:
Revert commit: ab98c7244215e0306bee2acdfe57c9cf48f3b8ff

willstocks committed:
Merge pull request #1 from willstocks-tech/Branch_ab98c72

Revert commit: ab98c7244215e0306bee2acdfe57c9cf48f3b8ff

jeherve commented:
The Unit tests are still failing: https://travis-ci.org/Automattic/jetpack/jobs/443692155#L661

You may have to run more tests: can you see the prefetch links on your test site? If so, you may have to update the unit tests to take the changes into account.

willstocks-tech commented:
Ah yes, I noticed that too @jeherve. I'll be honest, I'm not 100% sure how to update the tests to cater for these changes :(
I know it's to do with:
https://github.com/Automattic/jetpack/blob/2308333be3ef45a582314b6239cce91ca684aa1e/tests/php/test_class.jetpack.php#L209-L212
But I'm not quite sure how to reformat the `$expected` results. Any advice?

kraftbj commented:
This missing piece is this is hooked onto a filter, so the first parameter would need to be the resource hints being passed to it (not new_urls), and then we'd need to return them.

@jeherve Do you think it makes sense to do this via a static variable in the method? e.g. a static $prefetch; in the Jetpack class, and then have a new function to filter which just takes the prefetch variable and adds it to the value being filtered, or is there another approach?

willstocks-tech commented:
Don't think I'm fully understanding you @kraftbj - isn't `$new_urls` the array of urls that the `Jetpack::dns_prefetch` is collecting and passing back, therefore the hints that would be used?
I've just noticed, I could probably do with renaming `$urls` to `$hints` as that may make a bit more sense!

willstocks committed:
Wrap the core of the function in a single `if` statement, rather than multiple nested (repeated) `if`s
Also renamed `$urls` to `$hints` as this is a more appropriate name

willstocks committed:
Corrected dns_prefetch function indents - not sure what happened there!

willstocks committed:
Corrected dns_prefetch function indents - not sure what happened there!

kraftbj commented:
@willstocks-tech With a filter, `add_filter( 'wp_resource_hints', array( $this, 'dns_prefetch' ), 10, 2 );`, the Hooks API will call the `dns_prefetch` function and pass, in this case, two args per https://developer.wordpress.org/reference/hooks/wp_resource_hints/ -- the set of URLs WordPress has already and if they're doing prefetch or prerender.

The `dns-prefetch` function, in terms of being used as a filter, needs to pass back that first args with whatever changes we need to make (in this case additions).

A filter function without a return, in effect, does nothing -- something needs to get passed back by it.

But, to my question to jeherve, we're doing to do a lot more with this one function than I think is probably a good idea, so keep your changes to dns_prefetch and tweak things + a new function to actually serve out the URLs to the WP core filter.

willstocks committed:
Merge branch 'fix/dns-fetch-using-wp_resource_hints()' of https://github.com/willstocks-tech/jetpack into fix/dns-fetch-using-wp_resource_hints()

willstocks committed:
Correct typo

willstocks-tech commented:
Ahhh, OK I'm with you now @kraftbj!! 

At risk of being _wildly_ incorrect and looking a little stupid, could we not simply `return array($hints, $relation_type)` at the end of the function?

In the meantime, I'm hoping the latest commit I've pushed up is a cleaner, and more maintainable approach to the function, by wrapping the whole thing in a single `if` rather than having two of the same `if`s nested within? :)

kraftbj commented on class.jetpack.php:
```diff
@@ -6929,22 +6930,22 @@ public static function build_raw_urls( $url ) {
+	 * @param string | $relation_type | resource hint type
 	 */
-	public static function dns_prefetch( $new_urls = null ) {
+	public static function dns_prefetch( $new_urls = null, $relation_type = 'dns-prefetch' ) {
```
You're right wrt having a `return` to solve part of the problem, but something I missed in the beginning. For the filter, it has two arguments -- urls, and relationship. Our first two args of `dns_prefetch` needs to match that. They use `prefetch` instead of `dns-prefetch`, so that's one thing.

willstocks-tech replied:
```suggestion
	public static function dns_prefetch( $urls = null, $relation_type = 'dns-prefetch' ) {
```

kraftbj commented on class.jetpack.php:
```diff
@@ -6929,22 +6930,22 @@ public static function build_raw_urls( $url ) {
-			foreach ( (array) $new_urls as $this_new_url ) {
-				$prefetch_urls[] = strtolower( untrailingslashit( preg_replace( '#^https?://#i', '//', $this_new_url ) ) );
+		if( 'dns-prefetch' == $relation_type ) {
+			if ( empty( $new_urls ) && ! empty( $prefetch_urls ) ) {
```
Since this is on a filter, we can't use the lack of $new_urls as a way to check to see if we need to put them. We could do something like `doing_filter` to account for that and keep it all within one function.

I don't know if keeping it all within one function makes the most sense for long-term sustainability or if we're better off separating it out.

willstocks-tech replied:
OK, I _think_ I'm with you. If I'm understanding what you're suggesting:
1 x function to effectively capture the `$urls` (keep this function, pretty much as it is?) + 1 x function to capture the `$relation_type`, as opposed to trying to mash it all in to one "do it all" function?

willstocks-tech replied:
How do you feel about something like:
```
	if ( 'dns-prefetch' == $relation_type ) {
		foreach ( $urls as $key => $url ) {
			if ( strpos( $url, $prefetch_urls) == false ) {
				foreach ( (array) $new_urls as $this_new_hint ) {
					$hints[] = strtolower( untrailingslashit( preg_replace( '#^https?://#i', '//', $this_prefetch_hint ) ) );
				}
			}
		}
		$hints = array_unique( $hints );
	}
```

kraftbj reviewed (changes requested):
I don't want to send you on a goose chase, so let us get on the same page about how to complete.

I think using the same function is clever, but clever isn't always great. If we do, the logic map should be:
0. function takes two args, the url and the hint type
1. if doing the filter, the second arg is prefetch, and we have prefetch_urls, array merge the url var and return the url var.
2. else if there is a url, add it to the prefetch static url

willstocks-tech commented:
So @kraftbj - I've been doing a little research! What are your thoughts to the following "hashed up" version of Autoptimize's preconnect function: https://github.com/futtta/autoptimize/blob/736e79fb3d40d8ec679d9c668f22bc8e670892e3/classes/autoptimizeExtra.php#L302

`add_filter( 'wp_resource_hints', array( $this, 'dns_prefetch' ), 10, 2 );`
and:
```
    public function dns_prefetch( $hints, $relation_type ) {
         $prefetch_urls = array();
         // Get settings and store in array.
         $prefetch = $prefetch_urls;
         $prefetch = apply_filters( 'autoptimize_extra_filter_tobepreconn', $prefetch );
         // Walk array, extract domain and add to new array with crossorigin attribute.
         foreach ( $prefetch as $dnspre ) {
             $parsed = parse_url( $dnspre );
                    
             if ( is_array( $parsed ) && empty( $parsed['scheme'] ) ) {
                 $domain = '//' . $parsed['host'];
             } elseif ( is_array( $parsed ) ) {
                 $domain = $parsed['scheme'] . '://' . $parsed['host'];
             }
             if ( ! empty( $domain ) ) {
                 $hint = array( 'href' => $domain );
                 $new_hints[] = $hint;
             }
         }
         // Merge in WP's preconnect hints.
         if ( 'dns-prefetch' === $relation_type && ! empty( $new_hints ) ) {
             $hints = array_merge( $hints, $new_hints );
         }
         return $hints;
     }
```

This caters for the merging and handling of the wp included resource hints :)

willstocks-tech commented:
What do you think to the above @kraftbj / @jeherve ? I'm not 100% sure how to attack it otherwise, without fully splitting everything out into "separate" functions, as discussed at an earlier point

kraftbj commented:
Without testing, that direction makes sense to me. We'd probably want to `array_unique` before returning back to Core.

willstocks-tech commented:
Yes, that was my thought exactly too (I didn't mess with the function _too_ much). A simple: `$hints = array_unique( $hints );` before the return should do the trick.
I'll see if I can spin up a more appropriate and applicable function! :)

willstocks-tech commented:
Right - @jeherve / @kraftbj - how do you feel about this amended function, which I feel is much more appropriate!
```
//Based on Autoptimize's current preconnect function
//Should Jetpack instead utilize "preconnect" rather than dns-prefetch, as preconnect also takes care of SSL/TLS negotiation in advance
 public function filter_preconnect( $hints, $relation_type )
    {
        // Get prefetch urls and store in array.
        $preconns = array_filter( array_map( 'trim', explode( ',', $hints ) ) );
        $preconns = apply_filters( 'autoptimize_extra_filter_tobepreconn', $preconns );
		// Walk array, extract domain and add to new array with crossorigin attribute.
        foreach ( $preconns as $preconn ) {
            $parsed = parse_url( $preconn );
            if ( is_array( $parsed ) && empty( $parsed['scheme'] ) ) {
                $domain = '//' . $parsed['host'];
            } elseif ( is_array( $parsed ) ) {
                $domain = $parsed['scheme'] . '://' . $parsed['host'];
            }
            if ( ! empty( $domain ) ) {
                $hint = array( 'href' => $domain );
				// Can declare crossorigin attribute here if necessary (such as isrequired for fonts)
                $new_hints[] = $hint;
            }
        }
		// Merge in WP's standard preconnect hints.
        if ( 'preconnect' === $relation_type && ! empty( $new_hints ) ) {
            $hints = array_merge( $hints, $new_hints );
        }
		$hints = array_unique( $hints );
        return $hints;
    }
```
